zeebe/atomix/cluster/src/main/java/io/atomix/raft/impl/RaftContext.java

```diff
@@ -923,11 +935,30 @@ public long getFirstCommitIndex() {
    *
    * @param firstCommitIndex The first commit index.
    */
-  public void setFirstCommitIndex(final long firstCommitIndex) {
+  public void setFirstCommitIndex(final long firstCommitIndex, final long lastFlushedIndex) {
     if (this.firstCommitIndex == 0) {
       if (firstCommitIndex == 0) {
         return;
       }
+      // The previous leader may have crashed before updating the commit index on current leader.
+      // However, the data was already committed if the current leader had already flushed the
+      // message
+      // in its local storage.
+      // However, if the current leader has not received that message, we may be at risk of losing
+      // data.
```

Comment:
❓ I have some problems to follow this comment. Can we clarify this a bit.

Isn't the problem here that we had previous quorum of nodes, that replicated/committed data, now a new quorum either didn't receive that data or experienced a data loss? I'm talking here about quorum because it must be more then just the leader to run into this issue, plus they seem to be able to commit data.

Reply:
Yes, the actually is actually explaining why we need to check both `firstCommitIndex` and `lastFlushedIndex`.
In general this situation may arise (`L=Leader, F1=Follower 1, F2=Follower2`):
- L: appends entry `n` to follower
- F1,F2: append entry `n` to the log successfully
- L: tries sending an heartbeat (which contains the `commitIndex`) to the followers, but crashes.
This is the situation after the crash `(commitIndex, lastFlushedIndex)`:

| L | F1 | F2|
|-----| ---- | ---------
| (n,n) | (n-1, n) | (n-1, n)|

When F1 & F2 try to elect a leader, they will look at the lastFlushedIndex and agree that they should start from entry `n` as it's the latest entry they agree on, even though commitIndex is  `n -1` on both nodes